src/composite_systems.jl

```diff
@@ -0,0 +1,157 @@
+using LinearAlgebra: I
+import TensorOperations
+
+export CompositeQSystem, add_hamiltonian!, add_lindblad!
+
+# tensor products of quantum systems
+mutable struct CompositeQSystem
+    subsystems::Vector
+    fixed_Hs::Vector{Tuple} # tuple of Matrix and expansion indices
+    parametric_Hs::Vector{Tuple} # tuple of Functions and expansion indices
+    fixed_Ls::Vector{Tuple} # tuple of Matrix and expansion indices for collapse operators
+    parametric_Ls::Vector{Tuple} # tuple of Matrix, expansion indices, and time-dependent function for collapse operators
+    dimension::Int
+end
+
+CompositeQSystem(qs) = CompositeQSystem(qs, [], [], [], [], prod(dimension(q) for q in qs))
+
+# helper functions for CompositeQSystems
+dimension(cqs::CompositeQSystem) = cqs.dimension
+dim(cqs::CompositeQSystem) = dimension(cqs)
+
+find_indices(cqs::CompositeQSystem, s::QSystem) = findall([sub == s for sub in cqs.subsystems])
+find_indices(cqs::CompositeQSystem, s_label::AbstractString) = findall([label(sub) == s_label for sub in cqs.subsystems])
+function find_indices(cqs::CompositeQSystem, s::Union{Vector{<:QSystem}, Vector{<:AbstractString}})
+    return vcat([find_indices(cqs, _s) for _s in s]...)
+end
+
+""" Add a subsystem static Hamiltonian matrix to a CompositeQSystem """
+function add_hamiltonian!(cqs::CompositeQSystem, ham::AbstractMatrix{T}, acting_on::Union{Q, Array{Q}}) where {T<:Number, Q<:QSystem}
+    idxs = embed_indices(cqs, acting_on)
+    push!(cqs.fixed_Hs, (ham, idxs))
+end
+
+""" Add a subystem Hamiltonian to a CompositeQSystem """
+add_hamiltonian!(cqs::CompositeQSystem, qs::QSystem) = add_hamiltonian!(cqs, hamiltonian(qs), qs)
+add_hamiltonian!(cqs::CompositeQSystem, qs::QSystem, ϕ::Real) = add_hamiltonian!(cqs, hamiltonian(qs, ϕ), qs)
+
+# TODO how to do this dispatch vs adding a fixed Hamiltonian - Jameson says not to do this https://discourse.julialang.org/t/functions-and-callable-methods/2983/3
```

Comment:
I'm not sure how that comment from Jameson applies.

Reply:
Maybe I didn't understand the discussion there but I was looking to dispatch on whether an input argument was callable or not and I took Jameson's answer as affirming that it was not a good idea.